### ISSUE TITLE
add a new context manager to wrap `scipy.special.seterr`

### ISSUE THREAD

**Issue by ev-br** [scipy/scipy#6681]
TIL that `scipy.special.errprint` cannot be used as a context manager

```
In [16]: with errprint(True):
   ....:     print hyp2f1(0.5, 2/3., 1.5, 0.09j+0.75j)
   ....:     
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-16-c476ba18ca62> in <module>()
----> 1 with errprint(True):
      2     print hyp2f1(0.5, 2/3., 1.5, 0.09j+0.75j)
      3 

AttributeError: __exit__
```

Granted, it's not documented as such, so I guess here's an enhancement request to have a context manager similar to `numpy.errstate`, https://docs.scipy.org/doc/numpy/reference/generated/numpy.errstate.html

**Comment by person142** [comment]
I thought about doing this for a while, but I'm actually -1 on the idea. After doing `special.errprint(True)` the warnings can be controlled with the usual `warnings` framework.

**Comment by person142** [comment]
In fact, IMO the `errprint` function is only still useful because if you do `special.errprint(False)` it skips all the code that requires grabbing the GIL.

**Comment by ev-br** [comment]
Ugh, I see. So, to get an error, one needs to `>>> errstate(True)`, followed by

```
In [23]: with warnings.catch_warnings():
    warnings.simplefilter('error', SpecialFunctionWarning)
    print hyp2f1(0.5, 2/3., 1.5, 0.09j+0.75j)
   ....:     
---------------------------------------------------------------------------
SpecialFunctionWarning                    Traceback (most recent call last)
<ipython-input-23-ed30ca36cf79> in <module>()
      1 with warnings.catch_warnings():
      2     warnings.simplefilter('error', SpecialFunctionWarning)
----> 3     print hyp2f1(0.5, 2/3., 1.5, 0.09j+0.75j)
      4 

SpecialFunctionWarning: scipy.special/chyp2f1: loss of precision
```

(to write this I had to look up an exact warning class to filter out, then of course I forgot to import it). And I guess I typed this sort of code a few times already, so I get that easier than an average user.

On a constructive note, I think there are two things here: one is documentation and another one is a request for some syntax sugar.

For the docs: it's worth it to spell out the incantation either in the `scipy.special` tutorial, or even at the top of https://docs.scipy.org/doc/scipy/reference/special.html, under *Error handling*.

> In fact, IMO the errprint function is only still useful because if you do special.errprint(False) it skips all the code that requires grabbing the GIL.

And this, too!
Thanks for pointing this out, this should definitely be mentioned in the docs. (I did not know it.)

A separate bit of syntax sugar is adding a separate context manager similar to numpy's `seterr` and `errstate`. From numpy docs:

> Using an instance of errstate as a context manager allows statements in that context to execute with a known error handling behavior. Upon entering the context the error handling is set with seterr and seterrcall, and upon exiting it is reset to what it was before.

Likewise, a new context manager could defer to `scipy.special.errstate` upon entering, and restore the state on exit.

**Comment by pv** [comment]
I would be OK with a context manager, because every call to `errprint(True)` should be paired with `errprint(False)` because it's global state.

A separate issue is that errprint is iirc not thread-local.

**Comment by person142** [comment]
I can see the advantages of having a `seterr`/`errstate` pair; it would be nice to be able to e.g. easily ignore underflow while raising on loss of precision. If we add such functions I would push for doing away with `errprint` entirely so that there's one right way to do things.

I guess the right way to do this would be to map each `sf_error` category to a global variable which controls how we want that category to be handled (ignore, warn, raise, ...), and if it's ignore we avoid grabbing the GIL.

**Comment by person142** [comment]
@ev-br I wonder if you could comment on whether

https://github.com/person142/scipy/tree/special-seterr

fits the ticket? With that branch built here's how your session could have gone:

```
>>> import scipy.special as sc
>>> sc.geterr()
{'loss': 'ignore', 'arg': 'ignore', 'singular': 'ignore', 'no_result': 'ignore', 'other': 'ignore', 'overflow': 'ignore', 'slow': 'ignore', 'underflow': 'ignore', 'domain': 'ignore'}
>>> with sc.errstate(loss='raise'):
...     sc.hyp2f1(0.5, 2.0/3, 1.5, 0.09j+0.75j)
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
scipy.special.sf_error.SpecialFunctionError: scipy.special/chyp2f1: loss of precision
>>> sc.geterr()
{'loss': 'ignore', 'arg': 'ignore', 'singular': 'ignore', 'no_result': 'ignore', 'other': 'ignore', 'overflow': 'ignore', 'slow': 'ignore', 'underflow': 'ignore', 'domain': 'ignore'}
>>> with sc.errstate(loss='warn'):
...     sc.hyp2f1(0.5, 2.0/3, 1.5, 0.09j+0.75j)
...
__main__:2: SpecialFunctionError: scipy.special/chyp2f1: loss of precision
(0.93934867949609357+0.15593972567482395j)
>>> sc.geterr()
{'loss': 'ignore', 'arg': 'ignore', 'singular': 'ignore', 'no_result': 'ignore', 'other': 'ignore', 'overflow': 'ignore', 'slow': 'ignore', 'underflow': 'ignore', 'domain': 'ignore'}
```

It's still in a rudimentary state though (safety checks, tests, and documentation still to be added).

**Comment by person142** [comment]
Hooking this stuff up for the ufuncs is fine, but raising on the Cython bindings is a problem. In particular they're all `nogil`, so when an exception is raised in `sf_error` there's no way to propagate it. The usual mechanism would be using Cython's `except` to call `PyErr_Occurred` if there's a bad return value, but (1) IIRC it's broken for complex return values and (2) there isn't a consistent return value that indicates an error anyway (`nan` would be a good one, but something like `SF_ERROR_LOSS` could be an error but not return a `nan`). The only choice would be `except *` (modulo being broken for complex), but that defeats the purpose of using the bindings.

Possible solution: only allow warning? (Only allow warning for just Cython bindings?) I don't like this...

Is there something better?

**Comment by ev-br** [comment]
So the issue with cython_special is that e.g.

```
%%cython

cdef inner(double x) nogil:
    with gil:
        raise ValueError('oops')
    return x


def wrapper(double x):
    return inner(x)

wrapper(1.0)
```

does *not* raise `ValueError`, correct?

I can see an argument for ignoring these errors in cython wrappers altogether. The logic would then be 
that one would use ufuncs during debugging and only switch to nogil cython wrappers in a release build. 

Alternatively, I can also see an argument for making each cython wrapper have a additional pointer return of `int *errno`. (E.g., the signature for `cython_special.beta` would become `double beta(double, double, int *errno)` etc).

**Comment by pv** [comment]
That, or make `sf_error` error handling thread-local (which probably 
should be done in any case), and maintain an "sf_get_errno" interface 
for getting the latest error code.

**Comment by person142** [comment]
> That, or make `sf_error` error handling thread-local (which probably should be done in any case), and maintain an "sf_get_errno" interface for getting the latest error code.

Actually started this over the weekend. :-) Means we have to start looking for openmp at compile time though (not fun). :-(

> Alternatively, I can also see an argument for making each cython wrapper have a additional pointer return of `int *errno`.

In theory I like this better, but doing it would require touching every function in special...

**Comment by pv** [comment]
Do you actually need openmp? gcc has intrinsics for threadlocals, and in 
other cases we can fall back to Python, cf. gh-6509

**Comment by person142** [comment]
By fall back do you mean that if we can't get thread local storage then we use `PyGILState_Ensure()` to protect the global variable?

(If I'm understanding that correctly) I think openmp might be the most portable way to do it--should be supported on icc, MSVC, gcc, and clang >= 3.8. Target use case for `cython_special` is code using `cython.parallel`, which already uses openmp under the hood.

**Comment by pv** [comment]
Yes, if compiler doesn't provide TLS support, fall back to Python. Most 
compilers probably do provide it (it's included also in C11).
.
I'm not sure there's guarantee that openmp threads correspond to OS threads.

**Comment by person142** [comment]
> I'm not sure there's guarantee that openmp threads correspond to OS threads.

Yikes. Then if we make the error state thread local using compiler directives something like this might still be screwy:

```cython
from cython cimport parallel
cimport scipy.special.cython_special as sc

cdef int n
for n in parallel.prange(10, nogil=True):
    sc.gamma(<double>n)
    if sc.get_errno() == sc.SINGULAR:
        # do something
```

ISTM that if we want to use `cython_special` with error handling + `cython.parallel` we're essentially forced to use openmp and tell users to not use other kinds of threading (unless they agree to put no stock in error messages).

To be clear, openmp definitely isn't needed (and shouldn't be used) for the ufuncs. Just squabbling about the Cython side because I want things to play nicely with `cython.parallel`. (Though there's a non-negligible chance that I'm misunderstanding something somewhere, so corrections to my perspective are welcome.)

(Aside: the difficulties here are making me think of

> I can see an argument for ignoring these errors in cython wrappers altogether.

more and more fondly.)
